File: savethecat/savethecat.cpp
#include<iostream>
#include<iomanip>
#include<fstream>
#include<string>

using namespace std;

struct square{
	char c, move;
	int time=-1, t=-1, pos[2];
	bool checked = false, visited = false;
	square *previous = nullptr;
} map[1000][1000];

struct node {
	square *info;
	node *next;
};

class queue{
	public:
		queue(){
			first = nullptr;
			last = nullptr;
		}
		void push(square &s){
			node *temp = new node;
			temp->info = &s;
			temp->next = nullptr;
			if (first == nullptr) first = temp;
			else last->next = temp;
			last = temp;
		}
		square *front(){
			if (first != NULL) return first->info;
			else return NULL;
		}
		void pop(){
			node *temp = new node;
			temp = first;
			first = first->next;
			delete temp;
		}
		bool empty(){
			return first==nullptr;
		}
		void floodfill(int N, int M){
		   	while(!this->empty()){
				if (first->info->pos[0] -1 >= 0){
					if (map[first->info->pos[0]-1][first->info->pos[1]].c != 'X' && map[first->info->pos[0]-1][first->info->pos[1]].checked == false){
						map[first->info->pos[0]-1][first->info->pos[1]].time = first->info->time + 1;
						map[first->info->pos[0]-1][first->info->pos[1]].checked = true;
						this->push(map[first->info->pos[0]-1][first->info->pos[1]]);
					}
				}
				if (first->info->pos[0] +1 < N){
					if (map[first->info->pos[0]+1][first->info->pos[1]].c != 'X' && map[first->info->pos[0]+1][first->info->pos[1]].checked == false){
						map[first->info->pos[0]+1][first->info->pos[1]].time = first->info->time + 1;
						map[first->info->pos[0]+1][first->info->pos[1]].checked = true; 
						this->push(map[first->info->pos[0]+1][first->info->pos[1]]);
					}
				}
				if (first->info->pos[1] - 1 >=0){
					if (map[first->info->pos[0]][first->info->pos[1]-1].c != 'X' && map[first->info->pos[0]][first->info->pos[1]-1].checked == false){
						map[first->info->pos[0]][first->info->pos[1]-1].time = first->info->time + 1; 
						map[first->info->pos[0]][first->info->pos[1]-1].checked = true;
						this->push(map[first->info->pos[0]][first->info->pos[1]-1]);
					}
				}
				if (first->info->pos[1] + 1 < M){
					if (map[first->info->pos[0]][first->info->pos[1]+1].c != 'X' && map[first->info->pos[0]][first->info->pos[1]+1].checked == false){
						map[first->info->pos[0]][first->info->pos[1]+1].time = first->info->time + 1; 
						map[first->info->pos[0]][first->info->pos[1]+1].checked = true;
						this->push(map[first->info->pos[0]][first->info->pos[1]+1]);
					}
				}
				this->pop();
			}	
		}
		void catpath(int N, int M){
			int maxtime = 0;
			string path = "";
			square *maxsquare = new square;
			maxsquare = first->info;
			if (maxsquare->time != -1){
				while(! this->empty()){
					if (first->info->pos[0] + 1 < N){
						if (map[first->info->pos[0]+1][first->info->pos[1]].c != 'X' && map[first->info->pos[0]+1][first->info->pos[1]].visited == false && first->info->t+1 < map[first->info->pos[0]+1][first->info->pos[1]].time){
							map[first->info->pos[0]+1][first->info->pos[1]].t = first->info->t + 1;
							map[first->info->pos[0]+1][first->info->pos[1]].visited = true; 
							map[first->info->pos[0]+1][first->info->pos[1]].move = 'D';
							map[first->info->pos[0]+1][first->info->pos[1]].previous = new square;
							map[first->info->pos[0]+1][first->info->pos[1]].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (map[first->info->pos[0]+1][first->info->pos[1]].time - 1 > maxtime) {
								maxtime = map[first->info->pos[0]+1][first->info->pos[1]].time -1;
								maxsquare = &map[first->info->pos[0]+1][first->info->pos[1]];
							}
							else if ((map[first->info->pos[0]+1][first->info->pos[1]].time - 1 == maxtime) && (first->info->pos[0]+1 < maxsquare->pos[0])) maxsquare = &map[first->info->pos[0]+1][first->info->pos[1]];
							else if ((map[first->info->pos[0]+1][first->info->pos[1]].time - 1 == maxtime) && (first->info->pos[0]+1 == maxsquare->pos[0]) && (first->info->pos[1] < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]+1][first->info->pos[1]];
							this->push(map[first->info->pos[0]+1][first->info->pos[1]]);
						}
					}
					if (first->info->pos[1] - 1 >= 0){
						if (map[first->info->pos[0]][first->info->pos[1]-1].c != 'X' && map[first->info->pos[0]][first->info->pos[1]-1].visited == false && first->info->t+1 < map[first->info->pos[0]][first->info->pos[1]-1].time){
							map[first->info->pos[0]][first->info->pos[1]-1].t = first->info->t + 1; 
							map[first->info->pos[0]][first->info->pos[1]-1].visited = true;
							map[first->info->pos[0]][first->info->pos[1]-1].move = 'L';
							map[first->info->pos[0]][first->info->pos[1]-1].previous = new square;
							map[first->info->pos[0]][first->info->pos[1]-1].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (map[first->info->pos[0]][first->info->pos[1]-1].time - 1 > maxtime) {
								maxtime = map[first->info->pos[0]][first->info->pos[1]-1].time -1;
								maxsquare = &map[first->info->pos[0]][first->info->pos[1]-1];
							}
							else if ((map[first->info->pos[0]][first->info->pos[1]-1].time - 1 == maxtime) && (first->info->pos[0] < maxsquare->pos[0])) maxsquare = &map[first->info->pos[0]][first->info->pos[1]-1];
							else if ((map[first->info->pos[0]][first->info->pos[1]-1].time - 1 == maxtime) && (first->info->pos[0] == maxsquare->pos[0]) && (first->info->pos[1]-1 < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]][first->info->pos[1]-1];
							this->push(map[first->info->pos[0]][first->info->pos[1]-1]);
						}
					}
					if (first->info->pos[1] + 1 < M){
						if (map[first->info->pos[0]][first->info->pos[1]+1].c != 'X' && map[first->info->pos[0]][first->info->pos[1]+1].visited == false && first->info->t+1 < map[first->info->pos[0]][first->info->pos[1]+1].time){
							map[first->info->pos[0]][first->info->pos[1]+1].t = first->info->t + 1; 
							map[first->info->pos[0]][first->info->pos[1]+1].visited = true;
							map[first->info->pos[0]][first->info->pos[1]+1].move = 'R';
							map[first->info->pos[0]][first->info->pos[1]+1].previous = new square;
							map[first->info->pos[0]][first->info->pos[1]+1].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (map[first->info->pos[0]][first->info->pos[1]+1].time - 1 > maxtime) {
								maxtime = map[first->info->pos[0]][first->info->pos[1]+1].time -1;
								maxsquare = &map[first->info->pos[0]][first->info->pos[1]+1];
							}
							else if ((map[first->info->pos[0]][first->info->pos[1]+1].time - 1 == maxtime) && (first->info->pos[0] < maxsquare->pos[0])) maxsquare = &map[first->info->pos[0]][first->info->pos[1]+1];
							else if ((map[first->info->pos[0]][first->info->pos[1]+1].time - 1 == maxtime) && (first->info->pos[0] == maxsquare->pos[0]) && (first->info->pos[1]+1 < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]][first->info->pos[1]+1];
							this->push(map[first->info->pos[0]][first->info->pos[1]+1]);
						}
					}
					if (first->info->pos[0] - 1 >= 0){
						if (map[first->info->pos[0]-1][first->info->pos[1]].c != 'X' && map[first->info->pos[0]-1][first->info->pos[1]].visited == false && first->info->t+1 < map[first->info->pos[0]-1][first->info->pos[1]].time){
							map[first->info->pos[0]-1][first->info->pos[1]].t = first->info->t + 1;
							map[first->info->pos[0]-1][first->info->pos[1]].visited = true;
							map[first->info->pos[0]-1][first->info->pos[1]].move = 'U';
							map[first->info->pos[0]-1][first->info->pos[1]].previous = new square;
							map[first->info->pos[0]-1][first->info->pos[1]].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (map[first->info->pos[0]-1][first->info->pos[1]].time - 1 > maxtime) {
								maxtime = map[first->info->pos[0]-1][first->info->pos[1]].time -1;
								maxsquare = &map[first->info->pos[0]-1][first->info->pos[1]];
							}
							else if ((map[first->info->pos[0]-1][first->info->pos[1]].time - 1 == maxtime) && (first->info->pos[0]-1 < maxsquare->pos[0])) maxsquare = &map[first->info->pos[0]-1][first->info->pos[1]];
							else if ((map[first->info->pos[0]-1][first->info->pos[1]].time - 1 == maxtime) && (first->info->pos[0]-1 == maxsquare->pos[0]) && (first->info->pos[1] < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]-1][first->info->pos[1]];
							this->push(map[first->info->pos[0]-1][first->info->pos[1]]);
						}
					}
					this->pop();
				}
				cout << maxtime << endl;
				square *temp = new square;
				temp = maxsquare;
				while (temp->previous != nullptr) {
					path = temp->move + path;
					temp = temp->previous;
				}
				if ( path == "") cout << "stay\n";
				else cout << path << endl;
			}
			else{
				while(! this->empty()){
					if (first->info->pos[0] + 1 < N){
						if (map[first->info->pos[0]+1][first->info->pos[1]].c != 'X' && map[first->info->pos[0]+1][first->info->pos[1]].visited == false){
							map[first->info->pos[0]+1][first->info->pos[1]].visited = true; 
							map[first->info->pos[0]+1][first->info->pos[1]].move = 'D';
							map[first->info->pos[0]+1][first->info->pos[1]].previous = new square;
							map[first->info->pos[0]+1][first->info->pos[1]].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (first->info->pos[0]+1 < maxsquare->pos[0]) maxsquare = &map[first->info->pos[0]+1][first->info->pos[1]];
							else if ((first->info->pos[0]+1 == maxsquare->pos[0]) && (first->info->pos[1] < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]+1][first->info->pos[1]];
							this->push(map[first->info->pos[0]+1][first->info->pos[1]]);
						}
					}
					if (first->info->pos[1] - 1 >= 0){
						if (map[first->info->pos[0]][first->info->pos[1]-1].c != 'X' && map[first->info->pos[0]][first->info->pos[1]-1].visited == false){
							map[first->info->pos[0]][first->info->pos[1]-1].visited = true;
							map[first->info->pos[0]][first->info->pos[1]-1].move = 'L';
							map[first->info->pos[0]][first->info->pos[1]-1].previous = new square;
							map[first->info->pos[0]][first->info->pos[1]-1].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (first->info->pos[0] < maxsquare->pos[0]) maxsquare = &map[first->info->pos[0]][first->info->pos[1]-1];
							else if ((first->info->pos[0] == maxsquare->pos[0]) && (first->info->pos[1]-1 < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]][first->info->pos[1]-1];
							this->push(map[first->info->pos[0]][first->info->pos[1]-1]);
						}
					}
					if (first->info->pos[1] + 1 < M){
						if (map[first->info->pos[0]][first->info->pos[1]+1].c != 'X' && map[first->info->pos[0]][first->info->pos[1]+1].visited == false){
							map[first->info->pos[0]][first->info->pos[1]+1].visited = true;
							map[first->info->pos[0]][first->info->pos[1]+1].move = 'R';
							map[first->info->pos[0]][first->info->pos[1]+1].previous = new square;
							map[first->info->pos[0]][first->info->pos[1]+1].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (first->info->pos[0] < maxsquare->pos[0]) maxsquare = &map[first->info->pos[0]][first->info->pos[1]+1];
							else if ((first->info->pos[0] == maxsquare->pos[0]) && (first->info->pos[1]+1 < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]][first->info->pos[1]+1];
							this->push(map[first->info->pos[0]][first->info->pos[1]+1]);
						}
					}
					if (first->info->pos[0] - 1 >= 0){
						if (map[first->info->pos[0]-1][first->info->pos[1]].c != 'X' && map[first->info->pos[0]-1][first->info->pos[1]].visited == false){
							map[first->info->pos[0]-1][first->info->pos[1]].visited = true;
							map[first->info->pos[0]-1][first->info->pos[1]].move = 'U';
							map[first->info->pos[0]-1][first->info->pos[1]].previous = new square;
							map[first->info->pos[0]-1][first->info->pos[1]].previous = &map[first->info->pos[0]][first->info->pos[1]];
							if (first->info->pos[0]-1 < maxsquare->pos[0]) maxsquare = &map[first->info->pos[0]-1][first->info->pos[1]];
							else if ((first->info->pos[0]-1 == maxsquare->pos[0]) && (first->info->pos[1] < maxsquare->pos[1])) maxsquare = &map[first->info->pos[0]-1][first->info->pos[1]];
							this->push(map[first->info->pos[0]-1][first->info->pos[1]]);
						}
					}
					this->pop();
				}
				cout << "infinity" << endl;
				square *temp = new square;
				temp = maxsquare;
				while (temp->previous != nullptr) {
					path = temp->move + path;
					temp = temp->previous;
				}
				if ( path == "") cout << "stay\n";
				else cout << path << endl;
			}
		}
	private:
		node *first, *last;
};

int main(int argc, char **argv){
	ifstream infile;
	infile.open(argv[1]);
	char c;
	int N = 0, j=0, M = 0;
	queue flood, cat;	
	while(infile.get(c)){
		if(infile.eof()) break;
		if (c == '\n'){
		        N++;
			M = j;
			j=0;
		}
		else{
		       	map[N][j].c = c;
			map[N][j].pos[0] = N;
			map[N][j].pos[1] = j;
			if (c=='W') {
				map[N][j].time = 0;
				map[N][j].checked = true;
				flood.push(map[N][j]);
			}
			else if (c=='A'){
				map[N][j].t = 0;
				map[N][j].visited = true;
				cat.push(map[N][j]);
			}
			j++;
		}
	}
	flood.floodfill(N,M);
	cat.catpath(N, M);
}

File: savethecat/savethecat.py
import sys
import collections

d={}
flood = collections.deque()
cat = collections.deque()

def floodfill(N, M):
    while (flood):
        current = flood.popleft()
        if current[0] - 1 >= 0:
            if d[(current[0] - 1, current[1])][0] !='X' and d[(current[0]-1, current[1])][4] == False:
                d[(current[0] - 1, current[1])] = (d[(current[0] - 1, current[1])][0], '', d[current][2] + 1, d[(current[0] - 1, current[1])][3], True, d[(current[0] - 1, current[1])][5], None)
                flood.append((current[0]-1, current[1]))
        if current[0] + 1 < N:
    	    if d[(current[0] + 1, current[1])][0] !='X' and d[(current[0]+1, current[1])][4] == False:
                d[(current[0] + 1, current[1])] = (d[(current[0] + 1, current[1])][0], '', d[current][2] + 1, d[(current[0] + 1, current[1])][3], True, d[(current[0] + 1, current[1])][5], None)
                flood.append((current[0]+1, current[1]))
        if current[1] - 1 >= 0:
            if d[(current[0], current[1] - 1)][0] !='X' and d[(current[0], current[1] - 1)][4] == False:
                d[(current[0], current[1]-1)] = (d[(current[0], current[1]-1)][0], '', d[current][2] + 1, d[(current[0], current[1]-1)][3], True, d[(current[0], current[1]-1)][5], None)
                flood.append((current[0], current[1]-1))
        if current[1] + 1 < M:
            if d[(current[0], current[1] + 1)][0] !='X' and d[(current[0], current[1] + 1)][4] == False:
                d[(current[0], current[1]+1)] = (d[(current[0], current[1]+1)][0], '', d[current][2] + 1, d[(current[0], current[1]+1)][3], True, d[(current[0], current[1]+1)][5], None)
                flood.append((current[0], current[1]+1))

def catpath(N, M):
    maxtime = 0
    path = ""
    current = cat.popleft()
    maxsquare = current
    if d[maxsquare][2] != -1:
        while True:
            if current[0] + 1 < N:
                if d[(current[0] + 1, current[1])][0] !='X' and d[(current[0]+1, current[1])][5] == False and d[current][3]+1 < d[(current[0]+1, current[1])][2]:
                    d[(current[0]+1, current[1])] = (d[(current[0]+1, current[1])][0], 'D', d[(current[0]+1, current[1])][2], d[current][3]+1, True, True, current)
                    if d[(current[0]+1, current[1])][2] - 1 > maxtime:
                        maxtime = d[(current[0]+1, current[1])][2] - 1
                        maxsquare = (current[0]+1, current[1])
                    elif d[(current[0]+1, current[1])][2] - 1 == maxtime and (current[0]+1, current[1]) < maxsquare:
                        maxsquare = (current[0]+1, current[1])
                    cat.append((current[0]+1, current[1]))
            if current[1] - 1 >= 0:
                if d[(current[0], current[1]-1)][0] !='X' and d[(current[0], current[1]-1)][5] == False and d[current][3]+1 < d[(current[0], current[1]-1)][2]:
                    d[(current[0], current[1]-1)] = (d[(current[0], current[1]-1)][0], 'L', d[(current[0], current[1]-1)][2], d[current][3]+1, True, True, current)
                    if d[(current[0], current[1]-1)][2] - 1 > maxtime:
                        maxtime = d[(current[0], current[1]-1)][2] - 1
                        maxsquare = (current[0], current[1]-1)
                    elif d[(current[0], current[1]-1)][2] - 1 == maxtime and (current[0], current[1]-1) < maxsquare:
                        maxsquare = (current[0], current[1]-1)
                    cat.append((current[0], current[1]-1))
            if current[1] + 1 < M:
                if d[(current[0], current[1]+1)][0] !='X' and d[(current[0], current[1]+1)][5] == False and d[current][3]+1 < d[(current[0], current[1]+1)][2]:
                    d[(current[0], current[1]+1)] = (d[(current[0], current[1]+1)][0], 'R', d[(current[0], current[1]+1)][2], d[current][3]+1, True, True, current)
                    if d[(current[0], current[1]+1)][2] - 1 > maxtime:
                        maxtime = d[(current[0], current[1]+1)][2] - 1
                        maxsquare = (current[0], current[1]+1)
                    elif d[(current[0], current[1]+1)][2] - 1 == maxtime and (current[0], current[1]+1) < maxsquare:
                        maxsquare = (current[0], current[1]+1)
                    cat.append((current[0], current[1]+1))
            if current[0] - 1 >= 0:
                if d[(current[0] - 1, current[1])][0] !='X' and d[(current[0]-1, current[1])][5] == False and d[current][3]+1 < d[(current[0]-1, current[1])][2]:
                    d[(current[0]-1, current[1])] = (d[(current[0]-1, current[1])][0], 'U', d[(current[0]-1, current[1])][2], d[current][3]+1, True, True, current)
                    if d[(current[0]-1, current[1])][2] - 1 > maxtime:
                        maxtime = d[(current[0]-1, current[1])][2] - 1
                        maxsquare = (current[0]-1, current[1])
                    elif d[(current[0]-1, current[1])][2] - 1 == maxtime and (current[0]-1, current[1]) < maxsquare:
                        maxsquare = (current[0]-1, current[1])
                    cat.append((current[0]-1, current[1]))
            if not(cat):
                break
            current = cat.popleft()
        print(maxtime)
        temp = maxsquare
        while d[temp][6] != None:
            path = d[temp][1] + path
            temp = d[temp][6]
        if path == "":
            print("stay")
        else:
            print(path)
    else:
        while True:
            if current[0] + 1 < N:
                if d[(current[0] + 1, current[1])][0] !='X' and d[(current[0]+1, current[1])][5] == False:
                    d[(current[0]+1, current[1])] = (d[(current[0]+1, current[1])][0], 'D', -1, -1, False, True, current)
                    if (current[0]+1, current[1]) < maxsquare:
                        maxsquare = (current[0]+1, current[1])
                    cat.append((current[0]+1, current[1]))
            if current[1] - 1 >= 0:
                if d[(current[0], current[1]-1)][0] !='X' and d[(current[0], current[1]-1)][5] == False:
                    d[(current[0], current[1]-1)] = (d[(current[0], current[1]-1)][0], 'L', -1, -1, False, True, current)
                    if (current[0], current[1]-1) < maxsquare:
                        maxsquare = (current[0], current[1]-1)
                    cat.append((current[0], current[1]-1))
            if current[1] + 1 < M:
                if d[(current[0], current[1]+1)][0] !='X' and d[(current[0], current[1]+1)][5] == False:
                    d[(current[0], current[1]+1)] = (d[(current[0], current[1]+1)][0], 'R', -1, -1, False, True, current)
                    if (current[0], current[1]+1) < maxsquare:
                        maxsquare = (current[0], current[1]+1)
                    cat.append((current[0], current[1]+1))
            if current[0] - 1 >= 0:
                if d[(current[0] - 1, current[1])][0] !='X' and d[(current[0]-1, current[1])][5] == False:
                    d[(current[0]-1, current[1])] = (d[(current[0]-1, current[1])][0], 'U', -1, -1, False, True, current)
                    if (current[0]-1, current[1]) < maxsquare:
                        maxsquare = (current[0]-1, current[1])
                    cat.append((current[0]-1, current[1]))
            if not(cat):
                break
            current = cat.popleft()
        print("infinity")
        temp = maxsquare
        while d[temp][6] != None:
            path = d[temp][1] + path
            temp = d[temp][6]
        if path == "":
            print("stay")
        else:
            print(path)


if len(sys.argv) > 1:
    i = 0
    j = 0
    with open(sys.argv[1], "rt") as f:
        while True:
            l = f.readline()
            if l == "":
                break
            while l[j] != "\n":
                if l[j] == 'W':
                    d[i, j] = (l[j], '', 0, -1, True, False, None)
                    flood.append((i, j))
                elif l[j] == 'A':
                    d[i, j] = (l[j], '', -1, 0, False, True, None)
                    cat.append((i, j))
                else:
                    d[i, j] = (l[j], '', -1, -1, False, False, None)
                j += 1
            M = j
            i += 1
            j = 0
        N = i
floodfill(N, M)
catpath(N, M)
